### ISSUE TITLE
Remote-execution testing (technical debt)

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#629]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/629)
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Sep 5, 2018, 16:37

## Background

Initial remote-execution implementation (BuildStream/buildstream!626) does not include tests for the REAPI execution flow. This would require to somehow mock a remote-execution service and especially:
* The ``Execution`` service, the main end-point for the client.
* The ``ActionCache`` service, once BuildStream/buildstream#628 has been resolve / implemented.

The test-suite already has an ``ArtifactShare`` class that should be enough to mock the ``ContentAddressableStorage`` service. 

## Task description

See comments below and also the CI pipelines being set-up on BuildGrid - https://gitlab.com/BuildGrid/buildgrid.hub.docker.com/issues/2

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Sep 5, 2018, 16:39

mentioned in merge request !626

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Sep 24, 2018, 15:01

With BuildStream/buildstream#630, the remote execution client test suite should simulate connection drops so that client side operation stream reopening would be tested. Connection drops can be simulated from an `ExecutionServicer` by calling `cancel()` on the `Execute()` context instance.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Sep 25, 2018, 11:20

With BuildStream/buildstream#668, the remote execution client test suite should simulate build service errors (server or bot error) and build execution errors (non-zero build command return code).

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Oct 15, 2018, 12:00

This should include a test for #708, also see !871

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Oct 15, 2018, 12:02

mentioned in merge request !871

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @raoul]](https://gitlab.com/raoul).hidalgocharman on Oct 25, 2018, 16:19

mentioned in merge request !900

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @toscalix]](https://gitlab.com/toscalix) on Dec 10, 2018, 09:07

marked this issue as related to #794

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Dec 10, 2018, 09:17

marked #794 as a duplicate of this issue

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Dec 10, 2018, 09:24

[[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) what do you need to implement this? We can provide easily enough a group of servers to setup a mini cluster if needed.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Dec 10, 2018, 10:52

[[Gitlab user @jjardon]](https://gitlab.com/jjardon), [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard), FWIW, while I agree it would be nice to run automated CI on a mini cluster, it would also be nice to have our existing test suite use remote execution where possible.

I.e. it would be nice to reuse the same test battery we have in `./setup.py test`, and run all the same tests under remote execution by spawning a buildgrid locally, such that there is a setup/teardown of the grid and have the grid is isolated to the test (i.e. without hardcoded addresses, ensuring that a test grid never collides with a grid that the developer might be running on their laptop in parallel to their tests).

This would provide very good test coverage, and allow easier use for developers running tests locally on their own laptops.

Ideally, a new decorator could be implemented in the testutils module which:
* Automatically parameterizes the test to run with and without a BuildGrid
* Skips the BuildGrid case if a suitable BuildGrid is not installed locally
* Spawns/Tears down a BuildGrid on the local host
* Configures the BuildStream under test to use the spawned BuildGrid
* This would only be possible to use with the `cli` and `cli_integration` fixtures from our testutils

I.e. it would be very nice if we could just add a decorator to every test we think *should* pass with remote execution enabled (maybe some things explicitly dont work, like shells, under remote execution), and perhaps avoid adding the decorator to tests which we think remote execution doesnt change anything for (in order to save on test execution time).

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jmacarthur]](https://gitlab.com/jmacarthur) on Jan 3, 2019, 10:00

assigned to [[Gitlab user @jmacarthur]](https://gitlab.com/jmacarthur)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jmacarthur]](https://gitlab.com/jmacarthur) on Jan 4, 2019, 17:02

My initial research/plan for remote execution testing is at https://wiki.gnome.org/Projects/BuildStream/RemoteExecutionTesting. This isn't meant to be a permanent home for this document, but will serve while we're at this early stage.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @raoul]](https://gitlab.com/raoul).hidalgocharman on Jan 9, 2019, 16:32

mentioned in merge request !1013

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jmacarthur]](https://gitlab.com/jmacarthur) on Jan 29, 2019, 17:35

The wiki page above has been updated, but I'm unable to do any further work on this issue due to being reassigned.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jmacarthur]](https://gitlab.com/jmacarthur) on Jan 29, 2019, 17:35

unassigned [[Gitlab user @jmacarthur]](https://gitlab.com/jmacarthur)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jennis]](https://gitlab.com/jennis) on Feb 21, 2019, 17:30

I've tried to replicate Jim's work (described here: https://wiki.gnome.org/Projects/BuildStream/RemoteExecutionTesting) and have been unsuccessful. It looks like there is a problem with the buildbox worker.

Firstly, buildbox-fuse should now be built with cmake (not meson) so I had to slightly alter Jim's diff which needs to be patched to [BuildGrid's Dockerfile](https://gitlab.com/BuildGrid/buildgrid/blob/master/Dockerfile) in order to create an image with the buildbox binary (as well as an image for the buildgrid server). Dockerfile here:

```dockerfile
FROM debian:buster

RUN apt-get update && apt-get install -y \
    python3 \
    python3-pip

# Point the path to where buildgrid gets installed
ENV PATH=$PATH:/root/.local/bin/

# Upgrade python modules
RUN python3 -m pip install --upgrade setuptools pip

# Use /app as the current working directory
WORKDIR /app

# Copy the repo contents (source, config files, etc) in the WORKDIR
COPY . .

# Install BuildGrid
RUN pip3 install --user --editable .

WORKDIR /app

# From the buildbox-fuse dockerfile
RUN apt-get update && apt-get install -y \
    fuse3 \
    bubblewrap \
    cmake \
    gcc \
    g++ \
    git  \
    grpc++ \
    libfuse3-dev \
    libssl-dev \
    pkg-config \
    uuid-dev \
    && apt-get clean

RUN git clone https://gitlab.com/BuildGrid/buildbox/buildbox-fuse.git /buildbox
WORKDIR /buildbox

RUN mkdir build && cd build && cmake .. && make && ./buildbox --help
RUN mv build/buildbox /root/.local/bin

WORKDIR /app

# Entry Point of the image (should get an additional argument from CMD, the path to the config file)
ENTRYPOINT ["bgd", "server", "start", "-vv"]

# Default config file (used if no CMD specified when running)
CMD ["tls.yml"]
```

Now, I've used the same configuration (tls.yml) as Jim, but just stripped away all the authentication stuff, and set `insecure-mode:` to true. tls.yml below:

```yaml
server:
  - !channel
    port: 50051
    insecure-mode: true

description: |
  A single default instance.

instances:
  - name: ''
    description: |
      The main server

    storages:
      - !disk-storage &main-storage
        path: !expand-path $HOME/cas

    services:
      - !action-cache &main-action
        storage: *main-storage
        max-cached-refs: 256
        allow-updates: true

      - !execution
        storage: *main-storage
        action-cache: *main-action

      - !cas
        storage: *main-storage

      - !bytestream
        storage: *main-storage
```

Following the rest of Jim's advice, I received the following output from the bgd server (with it's own CAS storage) and the worker.

Buildgrid server:
```bash
 $ sudo docker run --rm -i -p 50051:50051 --network buildgrid-net --name bgd_server buildgrid_server2
2019-02-21 17:07:06,968:[           buildgrid.server.instance][ INFO]: Adding insecure connection on [[::]:50051]
2019-02-21 17:07:14,194:[      buildgrid.server.bots.instance][ INFO]: Created bot session name=[/c839152d-1b3c-4332-97ae-3e0e46441827] with bot_id=[None.998f82bf7447]
2019-02-21 17:07:49,434:[          buildgrid.server.scheduler][ INFO]: Job queued: [986436a2-b9b3-4607-a335-af0a265354fc]
2019-02-21 17:07:49,434:[  buildgrid.server.execution.service][ INFO]: Operation [/0cfe2342-72e7-4894-a3ed-cdd6b0da442a] created for job [986436a2-b9b3-4607-a335-af0a265354fc]
2019-02-21 17:07:49,866:[          buildgrid.server.scheduler][ INFO]: Job scheduled to run: [986436a2-b9b3-4607-a335-af0a265354fc]
```

Buildbot worker:
```bash
 $ sudo docker run --entrypoint 'bgd' --network buildgrid-net buildgrid_server2 bot --remote http://bgd_server:50051 --remote-cas http://bgd_server:50051 buildbox  --fuse-dir=/root/fuse --local-cas=/root/cache
2019-02-21 17:07:49,900:[                buildgrid.bot.tenant][ERROR]: b''
```

We can see that the worker is assigned work (at 17:07:49,866) and immediately when it tried to complete this work there is an error (at 17:07:49,900) with a pretty unhelpful error message.

I wonder if anyone has any ideas?

I'm also unable to do any further work on this for now due to being assigned to optimization work

-----

As an aside, I tried setting this up to use a docker image for BuildStream's artifact cache, and I was able to successfully communicate with the BuildStream cache from the BuildGrid server, so perhaps if we have a test which tests remote execution in CI, we should use the buildstream/artifact-cache docker image for the storage. As opposed to 'local' cas storage on the BuildGrid server.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jmacarthur]](https://gitlab.com/jmacarthur) on Feb 21, 2019, 17:43

I think the `[ERROR]: b''` message appears because Buildgrid is only reporting the stdout of a subprocess, not stderr. If you add `-vvv` to your `bgd` command line, you should see stderr which will probably give you more of a clue as to the problem.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @raoul]](https://gitlab.com/raoul).hidalgocharman on Feb 22, 2019, 14:14

mentioned in merge request !1174

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Feb 28, 2019, 14:24

We've recently created a new repo. for Buildgrid Docker builds [here](https://gitlab.com/BuildGrid/buildgrid.hub.docker.com). The images are not published on Dokcer Hub yet (should be soon) but are accessible from the repo. registry [here](https://gitlab.com/BuildGrid/buildgrid.hub.docker.com/container_registry).

BuildGrid has not stable release for now and won't get one anytime soon I believe. However, we are producing nightly images that should contains everything needed for BuildStream:
* `registry.gitlab.com/buildgrid/buildgrid.hub.docker.com/buildgrid:nightly`

The repo. also contains a dedicated Docker Compose manifest for the BuildStream use case: [`Composefile.buildstream.yml`](https://gitlab.com/BuildGrid/buildgrid.hub.docker.com/blob/master/Composefile.buildstream.yml). That one should help setting up a CI pipeline with BuildStream + BuildGrid and run test builds.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Feb 28, 2019, 15:17

Some notes taken from a recent chat on this:

## Purpose and method

Ultimate goal is to test BuildStream's remote execution as part of the main CI pipelines on the BuildStream project's master branch. For all intents and purposes, this means also testing Buildbox and BuildGrid. For this we'd need stable versions of each project to use, and they won't be available in the near future. So for now, to work towards something that does a minimal test of remote execution, we should aim for using the following branches: BuildStream - someones' branch (ie an MR), BuildGrid master and BuildBox-fuse master. This would look like:

* Using Docker compose (that Martin has written) to spawn a Docker container each for BuildGrid and BuildBox - see [repo](https://gitlab.com/BuildGrid/buildgrid.hub.docker.com) and [MR](https://gitlab.com/BuildGrid/buildgrid.hub.docker.com/merge_requests/1)
* Pulling in the Docker images that are generated nightly
* Having this in a separate repo to the main BuildStream CI
* Having failed tests not set to block the main BuildStream CI pipeline but throw a warning instead

So we're proposing that all CI jobs in BuildStream point towards a triggered CI in another repo that uses their branch and tests it with BuildGrid and Buildbox (both at master). There's a good chance people won't want this in the main CI pipeline at all. To be discussed.

## Detailed notes on where the effort is up to so far

Martin's repo has the docker-compose manifest, if we 'build' this and then `docker-compose -f Composefile.buildstream.yml up --scale bots=2` we spawn the following containers:

#### A 'storage' container (port 50052)

This is the BuildGrid (with bb-fuse) docker image, which bind mounts a storage.conf configuration file (./buildgrid/config/) into the container at /etc/buildgrid/. Uses the 'backend' and 'host' networks

This is CAS and Action Cache

#### A 'controller' container  (port 50051)

This uses the BuildGrid (with bb-fuse) Docker image and bind mounts a controller.conf file (./buildgrid/config) to /etc/buildgrid/. Uses the 'backend' and 'host' network.s

This is the execution service.

#### A 'bot' container

Also uses the BuildGrid (with bb-fuse) Docker image.

This is the buildbox worker and MUST RUN AS PRIVILEGED (because of bwrap).

Note that 'host' because BuildStream should be communicating with storage (remote cache) and BG server (execution service) over the same network. 'back end' because the execution service can be communicate with the workers over some other network.

Now, to run these in GitLab's pipeline, we must use a runner that uses the docker in docker image as the executor (or maybe we can have our own runner with a shell executor, not sure yet).

So, we need BuildStream somewhere... Installing BuildStream in the Docker container might be problematic, so why not just use another docker container (spawned within the DinD container alongside the bot, storage and controller containers).

Locally, we pulled the buildstream/testuite-debian image, ran this (with -i, PRIVILEGED, on the host network, WITHOUT declaring ports, and bind mounting in an autotools project with the correct config), cloned and installed bst master (this could be anyones branch) and managed to remotely build hello.bst and store it remotely.

One thing that needs more thought it that we 'shelled' into this BuildStream container (docker exec -it <container_name> /bin/bash), clone bst master, installed it, changed into the autotools project directory (the one I mounted) and built it (and saw that remote execution worked): we'll need to figure out how this is to be done in the .gitlab-ci script (which is executing commands in the DinD container).

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Mar 5, 2019, 13:13

marked this issue as related to BuildGrid/buildgrid.hub.docker.com#2

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Mar 5, 2019, 13:14

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 6, 2019, 12:46

First attempt at running BuildStream master in a remote-execution context against BuildGrid in a CI job: [here](https://gitlab.com/BuildGrid/buildgrid.hub.docker.com/-/jobs/172597308)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jennis]](https://gitlab.com/jennis) on Mar 6, 2019, 17:35

BuildStream remote execution is now being tested in the buildgrid.hub.docker.com repo: https://gitlab.com/BuildGrid/buildgrid.hub.docker.com/merge_requests/2

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 11, 2019, 13:26

mentioned in merge request !1214

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 12, 2019, 18:24

created branch [`mablanch/629-test-trigger`](https://gitlab.com/BuildStream/buildstream/compare/master...mablanch%2F629-test-trigger) to address this issue

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 13, 2019, 10:36

mentioned in commit BuildGrid/buildgrid.hub.docker.com[[Gitlab user @24de77f73c17cd568bca6516d74e4fbf58a86ef8]](https://gitlab.com/24de77f73c17cd568bca6516d74e4fbf58a86ef8)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 13, 2019, 13:19

Example of a BuildStream pipeline triggering a downstream end-to-end remote execution test: [here](https://gitlab.com/BuildStream/buildstream/pipelines/51631485).

I'd be happy working on a proper MR if we are fine with integrating that to BuildStream CI.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 14, 2019, 17:18

created branch [`mablanch/629-remote-execution-test`](https://gitlab.com/BuildStream/buildstream/compare/master...mablanch%2F629-remote-execution-test) to address this issue

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Mar 14, 2019, 18:26

mentioned in merge request !1232

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 19, 2019, 15:12

mentioned in merge request !1239

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 19, 2019, 16:38

WiP patch series are here: BuildStream/buildstream!1239.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 19, 2019, 17:19

mentioned in commit 210737600bd172a101d62e1866b8ba19f49fe4cb

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 19, 2019, 17:19

mentioned in commit 757ee09eaea3e7229f3640d0df7c2f99d818659a

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 19, 2019, 17:19

mentioned in commit 2094352ee73c5d66cd1c7a35d334caf45193eb48

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 19, 2019, 17:25

assigned to [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 20, 2019, 16:47

mentioned in commit 0e14d119a59016b285cb64d82d9410c8633f5878

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 20, 2019, 16:47

mentioned in commit 5e6fa37ab0140f80301a942def031a90f8ad08bd

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 20, 2019, 16:47

mentioned in commit 1e410824ac75ab6ab86e79f2303dff8570870d1d

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 20, 2019, 16:47

mentioned in commit 83010680f8d0cc635c2b2fbae114eb6d793c2e65

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 21, 2019, 11:07

mentioned in commit b6484fb1edf29c31d781e4e4b728e99820ece3a6

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 21, 2019, 11:07

mentioned in commit 60cdfc98acebe92779b22f328dc0860608b75fef

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 21, 2019, 11:07

mentioned in commit 4e7c9056e8f8a785f6d3db85957ab8978b719ca9

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 21, 2019, 11:07

mentioned in commit 226cfeff5eae35b1b89a7a780c0e856179040e9c

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 21, 2019, 11:18

mentioned in commit cb5e358670b8ee8e9f12678bcf51caae10e5d17f

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 21, 2019, 11:18

mentioned in commit 6350f066b3186749e9b3232b210507e6c59a1e31

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 21, 2019, 11:18

mentioned in commit 2e4d9abc624ebf155e1315ed3133107887a76e89

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 22, 2019, 12:25

mentioned in commit 30eab915796ea0c4a41e486695a64aca9e7aa9b8

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 22, 2019, 12:25

mentioned in commit 99b63d716bc9bd5c54d32d12a792570ece9e024c

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 22, 2019, 12:25

mentioned in commit 015a98bda7f7aadbfc8b8c38bee546d7616e516d

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 22, 2019, 12:25

mentioned in commit fe63c5dea733c9b9d7dd2f4fb97ef53490a427a5

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 22, 2019, 12:25

mentioned in commit e32cddd1952ba81d52246ba1920481ee76b78687

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 22, 2019, 13:35

mentioned in commit 128ecfa21613837f2492dd5984614270be6d9aa8

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 22, 2019, 13:35

mentioned in commit e484716d80d19a609582672c58ba0564f194dfee

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 22, 2019, 13:35

mentioned in commit e2d6350d88071825266325d6d012fa673353b4a0

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 22, 2019, 13:35

mentioned in commit b56358c63c572f09aa0dd9d64f170054f0149c0c

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 22, 2019, 13:35

mentioned in commit a73d178dbd801b89100f5aef1608ff3771d504e4

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 25, 2019, 10:03

mentioned in commit 78aed2f1985630f9ff2be490d6da6e0be947492e

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 25, 2019, 10:03

mentioned in commit fa3bc8d12d32211d5f6cc85e0aab8ec7d90324c4

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 25, 2019, 10:03

mentioned in commit 839a734485339b848cf2e212984e366c0b399332

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 25, 2019, 10:03

mentioned in commit 561257dd706e82b93510c7ce0b3a229f01443e7b

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 25, 2019, 10:03

mentioned in commit 2623c6ee9dc9b3bfa632ba9b0b331aa34e8e66ba

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 25, 2019, 12:42

mentioned in commit 21b1dea923d01e3d2d4179e2189595aa9ae678e8

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 25, 2019, 12:42

mentioned in commit b626fa623367bd3f62f38c5cf0c8b8b2eaf222a4

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 25, 2019, 12:42

mentioned in commit 686a83e3faa9f6f9823c719b23c766fa11023a70

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 25, 2019, 12:42

mentioned in commit b0205501bdc99af1470baf7caad9e8f09f619aeb

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @martinblanchard]](https://gitlab.com/martinblanchard) on Mar 25, 2019, 12:42

mentioned in commit 36603fa8ed8ffd8eae5b8a3e84b7446cac24b6ac

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Mar 25, 2019, 16:49

Closing since https://gitlab.com/BuildStream/buildstream/merge_requests/1239 has been merged.

Thanks Martin!

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Mar 25, 2019, 16:49

closed
